src/Features/Core/Portable/MakeMethodSynchronous/AbstractMakeMethodSynchronousCodeFixProvider.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MakeMethodSynchronous
+{
+    internal abstract class AbstractMakeMethodSynchronousCodeFixProvider : CodeFixProvider
+    {
+        protected abstract bool IsMethodOrAnonymousFunction(SyntaxNode node);
+        protected abstract SyntaxNode RemoveAsyncTokenAndFixReturnType(IMethodSymbol methodSymbolOpt, SyntaxNode node);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var document = context.Document;
+            var diagnostic = context.Diagnostics.First();
+
+            var token = diagnostic.Location.FindToken(cancellationToken);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            for (var node = token.Parent; node != null; node = node.Parent)
+            {
+                if (IsMethodOrAnonymousFunction(node))
+                {
+                    context.RegisterCodeFix(new MyCodeAction(FeaturesResources.Make_method_synchronous,
+                        c => FixNodeAsync(context.Document, node, c)), diagnostic);
+                    return;
+                }
+            }
+        }
+
+        private const string AsyncSuffix = "Async";
+
+        private async Task<Solution> FixNodeAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var methodSymbolOpt = semanticModel.GetDeclaredSymbol(node) as IMethodSymbol;
+
+            // If the method ends with 'Async' then remove that as part of the fix.
+            if (methodSymbolOpt?.MethodKind == MethodKind.Ordinary && 
+                methodSymbolOpt.Name.Length > AsyncSuffix.Length && 
```

Comment:
EndsWith handles this (or is this to handle the case where the method is simply named `Async`)?